### PR TITLE
Refactored EntityHandleManager and HandleHolderBase

### PR DESCRIPTION
**Pull request by csilvestrini**
An assortment of sort-of related clean ups:

* There were a scattering of mostly duplicated methods in EntityHandleManager that I simplified.
* Moved HandleHolder initialisation into AbstractArcHost.
* Updated the HandleHolder interface and implementation to be a bit more rigid.
* Added exception logging to AbstractArcHost.

### CODE REVIEW

**Comment by csilvestrini** [comment]
presubmit cl/299024006

File: java/arcs/core/host/AbstractArcHost.kt [hunk 1]
```diff
@@ -37,6 +36,7 @@ const val MAX_CONSECUTIVE_FAILURES = 5
  * @property initialParticles The initial set of [Particle]s that this host contains.
  */
 abstract class AbstractArcHost(vararg initialParticles: ParticleRegistration) : ArcHost {
+    private val log = TaggedLog { "ArcHost" }
```
Comment (cromwellian) [inline java/arcs/core/host/AbstractArcHost.kt]: Should this be AbstractArcHost?

File: java/arcs/core/host/AbstractArcHost.kt [hunk 2]
```diff
@@ -308,24 +315,24 @@ abstract class AbstractArcHost(vararg initialParticles: ParticleRegistration) :
...
                 handleSpec.type.toSchema()
             )
         else -> throw IllegalArgumentException("Unknown type ${handleSpec.type}")
-    }
+    }.also { holder.setHandle(handleName, it) }
 
     override suspend fun stopArc(partition: Plan.Partition) {
         val arcId = partition.arcId
-        val context = lookupOrCreateArcHostContext(partition)
+        lookupOrCreateArcHostContext(partition)
```
Comment (cromwellian) [inline java/arcs/core/host/AbstractArcHost.kt]: the lookupOrCreate line can be deleted. stopArc() can't possibly create a context, one must already exist, which is why the line below looks up the context via runningArcs[partition.arcId] and only executes if it's non null.

My guess is I was refactoring and that no-op line got left as I was deleting code.

File: java/arcs/core/host/AbstractArcHost.kt [hunk 3]
```diff
@@ -383,9 +391,7 @@ abstract class AbstractArcHost(vararg initialParticles: ParticleRegistration) :
      * Unregisters [Handle]s from [StorageProxy]s, and clears references to them from [Particle]s.
      */
     private suspend fun cleanupHandles(handles: HandleHolder) {
-        for ((name, handle) in handles.handles) {
-            // TODO: disconnect/unregister handle
-        }
+        // TODO: disconnect/unregister handles
```
Comment (cromwellian) [inline java/arcs/core/host/AbstractArcHost.kt]: This reminds me, we need to file an issue, as we need an api to tell handles to unregister themselves, and some way to tell the storageProxy/ActiveStore it is no longer needed and can close its resources.

File: java/arcs/sdk/Particle.kt [hunk 4]
```diff
@@ -24,9 +24,42 @@ typealias Particle = arcs.core.host.Particle
...
+        }
+        handles[handleName] = handle
+    }
+
     override fun clear() = handles.clear()
+
+    private fun checkHandleIsValid(handleName: String) {
+        if (!entitySpecs.containsKey(handleName)) {
+            throw NoSuchElementException(
+                "Particle $particleName does not have a handle with name $handleName."
```
Comment (cromwellian) [inline java/arcs/sdk/Particle.kt]: does not have an 'EntitySpec with name $handleName' ?

File: java/arcs/core/host/AbstractArcHost.kt (hunk 3 above)
Comment (jblebrun) [reply java/arcs/core/host/AbstractArcHost.kt]: I am working on handle start/stop API now, just created a ticket for it. 

I'll create a ticket for the storageProxy / store work as well.

File: java/arcs/core/host/AbstractArcHost.kt (hunk 1 above)
Comment (csilvestrini) [reply java/arcs/core/host/AbstractArcHost.kt]: Done

File: java/arcs/core/host/AbstractArcHost.kt (hunk 2 above)
Comment (csilvestrini) [reply java/arcs/core/host/AbstractArcHost.kt]: Done

File: java/arcs/sdk/Particle.kt (hunk 4 above)
Comment (csilvestrini) [reply java/arcs/sdk/Particle.kt]: This one was a little subtle. The entire `List<EntitySpec>` gets passed in when this object is created, so we can expect that it has a full list of handle names and is reasonably authoritative. Whereas the handle list can grow over time as handles get created, so we want to check handle names against the entitySpecs map to see if they are valid.

I'll add a comment.

**Pull request merged**